oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp

```diff
@@ -0,0 +1,280 @@
+#include "pfw_sms_menu.hpp"
+
+#include <security/pam_appl.h>
+
+#include <cstdlib>
+#include <cstring>
+#include <ranges>
+
+namespace ibm_pfw_sms
+{
+
+/** @struct Data for the PAM Conversation function
+ *
+ * @note This is passed from our client code through PAM into the conversation
+ * via appdata_ptr
+ *
+ * [in] password - this the password that our application
+ * already has before invoking PAM. For authentication, this is the current
+ * password. For password change, this is the proposed new password.
+ * [out] changePasswordReasonCode - If one of the PAM modules (such as
+ * pam_pwquality) gives a response, we convert it to a reason code and store it
+ * here. The caller will need examine both this code and the return value from
+ * the pam_* function.
+ */
+struct pfw_sms_pam_appdata
+{
+    const std::string& password;
+    enum changePasswordReasonCode reasonCode;
+};
+
+/** @brief Our PAM Conversation function
+ *
+ * This function handles the following elements of the PAM conversation:
+ *  - For both authentication and password change, when PAM asks for the
+ * password, we supply it.
+ *  - For the change password function, if a module (such as pam_pwquality.so)
+ * complains about the password quality, we translate that message into a reason
+ * code (appdata_ptr->reasonCode). We use struct pfw_sms_pam_appdata (as this
+ * function's appdata_ptr) to pass data to and from our application and PAM:
+ * @note For parameter descriptions, see docs for pam_start and pam_conv
+ * @return PAM status code (int)
+ */
+static int pamConversationFunction(int num_msg,
+                                   const struct pam_message** message_ptr,
+                                   struct pam_response** response_ptr,
+                                   void* appdata_ptr)
+{
+    if ((appdata_ptr == nullptr) || (message_ptr == nullptr) ||
+        (response_ptr == nullptr))
+    {
+        return PAM_CONV_ERR;
+    }
+    if (num_msg <= 0 || num_msg >= PAM_MAX_NUM_MSG)
+    {
+        return PAM_CONV_ERR;
```

Comment:
could you please  improve tracing in case of failure cases.

Reply:
Individual PAM modules (such as pam_pwquality.so) write messages to the journal; this is controlled by the /etc/pam.d/ files, using the QUIET argument.
I don't have any ideas for what more to trace at this layer.  In this particular line of code, num_msg==1 always, for historical reasons, so we are already being paranoid here, and never expect this error.